src/components/Filter/Filter.js

```diff
@@ -1,113 +1,116 @@
 import './Filter.css';
 
-import React, { Component } from 'react';
+import React, { useReducer } from 'react';
 import classNames from 'classnames';
 import AutoComplete from '../AutoComplete/AutoComplete';
 import Input from '../Input/Input';
 import Switch from '../Switch/Switch';
 
 import { generateLists } from '../../listsGenerator';
 
-export default class Filter extends Component {
-  state = {
-    tag: '',
-    country: '',
-    name: '',
-    language: '',
-    showFilters: false,
-  };
+const Filter = props => {
+  const [state, setState] = useReducer(
+    (state, newState) => ({ ...state, ...newState }),
+    {
+      tag: '',
+      country: '',
+      name: '',
+      language: '',
+      showFilters: false,
+    }
+  );
 
-  onTagSelect = tag => {
-    this.setState({ tag });
-    this.props.onTagSelected(tag);
+  const onTagSelect = tag => {
+    setState({ tag });
+    props.onTagSelected(tag);
```

Comment:
Is this function guaranteed to be passed down?  If not, we should probably guard against its usage like so:

```suggestion
    if (props.onTagSelected != null) {
      props.onTagSelected(tag);
    }
```

Reply:
I think so. Do you think we should add `propTypes` to enable the check in development?

Reply:
`propTypes` would be helpful, but they don't guarantee that a prop is present; they only give feedback if the prop is missing or misconfigured.  Using a guard is the only way to be 100% sure we don't create a runtime error.

Reply:
> Using a guard is the only way to be 100% sure we don't create a runtime error.

After giving this more thought, a `propType` would be fine if these event handlers (e.g. onTagSelected) are required.  If they are optional we should still guard.

Reply:
We can guard for now just to make sure. Once we decide what approach to take (`propType` , `TypeScript` etc.) we could remove that check..

Reply:
Cool. Will get to this in a while.